org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/ProjectsManager.java

```diff
@@ -180,6 +182,39 @@ public IStatus runInWorkspace(IProgressMonitor monitor) {
 		return job;
 	}
 
+	public void cleanupProjects() throws CoreException {
+		IProject proj = this.getDefaultProject();
+		IJavaProject javaProj = JavaCore.create(proj);
+
+		Arrays.stream(javaProj.getPackageFragments()).filter(packageFragment -> {
+			try {
+				return packageFragment.containsJavaResources() && packageFragment.getKind() == IPackageFragmentRoot.K_SOURCE;
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem getting package fragments when cleanup projects", e);
+			}
+			return false;
+		}).flatMap(packageFragment -> {
+			try {
+				return Arrays.stream(packageFragment.getCompilationUnits());
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem getting compilation units when cleanup projects", e);
+			}
+			return null;
+		}).forEach((cu) -> {
+			try {
+				IResource resource = cu.getResource();
+				if (resource.isLinked()) {
+					File f = new File(cu.getUnderlyingResource().getLocationURI());
+					if (!f.exists()) {
+						cu.delete(true, null);
+					}
+				}
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Problem deleting compilation unit when cleanup projects", e);
```

Comment:
"Unable to delete missing compilation unit ("+resource.getName()+") from "+project.getName()